lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -988,6 +988,155 @@ OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
   return {};
 }
 
+// Given a range of operands, computes the number of matching prefix and
+// suffix elements. This does not perform cross-element matching.
+template <typename Range>
+static size_t computeCommonPrefixLength(const Range &a, const Range &b) {
+  size_t commonPrefixLength = 0;
+  auto ia = a.begin();
+  auto ib = b.begin();
+
+  for (; ia != a.end() && ib != b.end(); ia++, ib++, commonPrefixLength++) {
+    if (*ia != *ib) {
+      break;
+    }
+  }
+
+  return commonPrefixLength;
+}
+
+static size_t getTotalWidth(const OperandRange &range) {
+  size_t totalWidth = 0;
+  for (auto operand : range) {
+    // getIntOrFloatBitWidth should never raise, since all arguments to ConcatOp
+    // are integers.
+    ssize_t width = operand.getType().getIntOrFloatBitWidth();
+    assert(width >= 0);
+    totalWidth += width;
+  }
+  return totalWidth;
+}
+
+static bool isAnyZeroWidthOperands(OperandRange operands) {
+  return llvm::any_of(operands, [&](auto x) {
+    return x.getType().getIntOrFloatBitWidth() == 0;
+  });
+}
+
+// Reduce the strength icmp(concat(...), concat(...)) by doing a element-wise
+// comparison on common prefix and suffixes. Returns success() if a rewriting
+// happens.
+static LogicalResult matchAndRewriteCompareConcat(ICmpOp op, ConcatOp lhs,
+                                                  ConcatOp rhs,
+                                                  PatternRewriter &rewriter) {
+
+  auto lhsOperands = lhs.getOperands();
+  auto rhsOperands = rhs.getOperands();
+
+  if (lhs.getType().getWidth() == 0 || rhs.getType().getWidth() == 0) {
+    return failure();
+  }
+
+  if (isAnyZeroWidthOperands(lhsOperands) ||
+      isAnyZeroWidthOperands(rhsOperands)) {
```

Comment:
Plz add some comments about why these are rejected (I know today but wont remember in a year!).  Also consider merging the two if conditions together.

Alternatively: Is the first if condition needed?  Is a zero operand concat valid?  I don't think it should be, the verifier should reject this.